File: golang/152_maximum_product_subarray.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func maxProduct(nums []int) int {
	tMax := nums[0]
	tMin := nums[0]
	ans := nums[0]
	for i := 1; i < len(nums); i++ {
		if nums[i] < 0 {
			temp := tMax
			tMax = tMin
			tMin = temp
		}
		if tMax*nums[i] > nums[i] {
			tMax = tMax * nums[i]
		} else {
			tMax = nums[i]
		}
		if tMin*nums[i] < nums[i] {
			tMin = tMin * nums[i]
		} else {
			tMin = nums[i]
		}
		if tMax > ans {
			ans = tMax
		}
	}
	return ans
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	line := strings.Split(scanner.Text(), ",")
	var nums []int
	for i := 0; i < len(line); i++ {
		num, err := strconv.Atoi(line[i])
		if err == nil {
			nums = append(nums, num)
		}
	}
	fmt.Println(maxProduct(nums))
}

File: python/152_maximum_product_subarray.py
from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        t_max = nums[0]
        t_min = nums[0]
        ans = nums[0]
        for i in range(1, len(nums)):
            if nums[i] < 0:
                temp = t_max
                t_max = t_min
                t_min = temp
            t_max = max( t_max * nums[i], nums[i])
            t_min = min( t_min * nums[i], nums[i])
            ans = max( ans, t_max)
        return ans

sol = Solution()
inp = [int(x) for x in input().split(",")]
print(sol.maxProduct(inp))